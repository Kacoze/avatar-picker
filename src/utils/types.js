import propTypes from 'prop-types';

const AvatarShape = propTypes.shape({
  src: propTypes.string.isRequired,
  label: propTypes.string.isRequired,
  id: propTypes.number.isRequired,
});

const AvatarListShape = propTypes.arrayOf(AvatarShape);

export { AvatarShape, AvatarListShape };
