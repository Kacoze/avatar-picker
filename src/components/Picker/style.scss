@import '~shared/styles/index.scss';
@import url('https://fonts.googleapis.com/css?family=Source+Sans+Pro');

$size: 60px;
$space: 14px;
$container: 282px;
$arrow-size: 8px;
$border-big: 3px;

.Picker {
  position: relative;
}
.avatarButton {
  @include circle($size);
  padding: 0;
  border: none;
  outline: 0;
}

.avatarButton:hover,
.avatarButton:focus {
  border: 1px solid $border-p;
}


.Popover {
  position: absolute;
  left: -$container / 2 + $space;
  top: $size + $space;
  background: $background;
  box-shadow: 2px 2px 10px $shadow;
  border-radius: 2px;
  padding: $space;
  width: $container;
  text-align: center;
}

.Popover::before {
  content: ' ';
  position: absolute;
	top: -$arrow-size;
	left: $container/2 + $arrow-size;
	border-left: $arrow-size solid transparent;
	border-right: $arrow-size solid transparent;
	border-bottom: $arrow-size solid $background;
  @include sizing(0);
}

.PickerHeader{
  font-family: 'Source Sans Pro';
  color: $white;
  size: $header;
  margin: $space auto;
}

.avatar {
  @include circle(100%);
  padding: 0;
  margin: 0;
}

.AvatarList {
  display: flex;
  flex-wrap: wrap;
  .avatarButton {
    z-index: $z-index-1;
    margin: $space $space $space 0;
    border: $border-big solid $background;
  }
  .avatarButton:nth-child(4n) {
    margin-right: 0;
  }
  .avatarButton:hover,
  .avatarButton:focus {
    border: $border-big solid $border-p-o2;
    .avatar {
      background: $action;
      opacity: $opacity;
    }
  }
  .loading {
    position: absolute;
    display: inline-block;
    @include circle($size);
    border: $border-big solid $background;
    animation: spin 0.75s infinite linear;
    margin-left: -$size;
    margin-top: -$border-big - $border-big;
  }

  .loadingButton {
    border: $border-big solid $white;
  }

  @keyframes spin {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(360deg);
    }
  }

  .loading,
  .loading::before,
  .loading::after {
    border-color: $action;
    border-top-color: transparent;
    animation-duration: 1.2s;
  }
}
