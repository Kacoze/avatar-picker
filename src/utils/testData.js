const AvatarListExample = [
  { src: 'images/avatar/avatar1.png', label: 'Avatar 1', id: 1 },
  { src: 'images/avatar/avatar2.png', label: 'Avatar 2', id: 2 },
  { src: 'images/avatar/avatar3.png', label: 'Avatar 3', id: 3 },
  { src: 'images/avatar/avatar4.png', label: 'Avatar 4', id: 4 },
  { src: 'images/avatar/avatar5.png', label: 'Avatar 5', id: 5 },
  { src: 'images/avatar/avatar6.png', label: 'Avatar 6', id: 6 },
];
const AvatarExample = AvatarListExample[0];

export { AvatarListExample, AvatarExample };
