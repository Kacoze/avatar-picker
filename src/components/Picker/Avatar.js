import React from 'react';
import propTypes from 'prop-types';
import { AvatarShape } from '../../utils/types';
import styles from './style.scss';

function Avatar({
  image,
  onAction,
  loading,
  id,
}) {
  return (
    <button
      className={`${styles.avatarButton} ${loading && styles.loadingButton}`}
      onClick={() => onAction(image.id)}
    >
      <img
        src={image.src}
        alt={image.label}
        className={styles.avatar}
        id={id}
      />
      {loading && <div className={styles.loading} />}
    </button>
  );
}

Avatar.propTypes = {
  image: AvatarShape.isRequired,
  onAction: propTypes.func.isRequired,
  loading: propTypes.bool,
  id: propTypes.string,
};

Avatar.defaultProps = {
  loading: false,
  id: null,
};


export default Avatar;
