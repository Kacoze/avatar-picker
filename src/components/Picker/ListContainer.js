import React, { Component } from 'react';
import propTypes from 'prop-types';
import { AvatarListShape } from '../../utils/types';
import Avatar from './Avatar';
import styles from './style.scss';

class ListContainer extends Component {
  handleClickOutside = () => this.props.handleActiveClick();
  render() {
    const {
      list,
      activeAvatar,
      loading,
      handleAvatarChange,
      handleActiveClick,
    } = this.props;
    return (
      <div className={styles.Popover}>
        <h4 className={styles.PickerHeader}>Choose your avatar</h4>
        <div className={styles.AvatarList}>
          <Avatar
            image={list[activeAvatar - 1]}
            onAction={handleActiveClick}
          />
          {list.map(image => image.id !== list[activeAvatar - 1].id && (
            <Avatar
              key={image.id}
              image={image}
              onAction={handleAvatarChange}
              loading={loading === image.id}
            />
          ))}
        </div>
      </div>
    );
  }
}

ListContainer.propTypes = {
  list: AvatarListShape.isRequired,
  loading: propTypes.oneOfType([propTypes.bool, propTypes.number]).isRequired,
  activeAvatar: propTypes.number.isRequired,
  handleAvatarChange: propTypes.func.isRequired,
  handleActiveClick: propTypes.func.isRequired,
};

export default ListContainer;
