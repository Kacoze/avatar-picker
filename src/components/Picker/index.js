import React, { Component } from 'react';
import Outside from 'react-click-outside';
import { AvatarListShape } from '../../utils/types';
import Avatar from './Avatar';
import styles from './style.scss';
import ListContainer from './ListContainer';

const EnhancedListContainer = Outside(ListContainer);

class Picker extends Component {
  state = {
    loading: false,
    activeAvatar: 1,
    visibleList: false,
  }
  request = cb => setTimeout(cb, 1000);
  handleActiveClick = () => this.setState({ visibleList: !this.state.visibleList });
  handleAvatarChange = (id) => {
    this.setState({ loading: id });
    this.request(() => this.setState({ activeAvatar: id, visibleList: false, loading: false }));
  }
  render() {
    const { list } = this.props;
    const { activeAvatar, visibleList, loading } = this.state;
    return (
      <div className={styles.Picker}>
        <Avatar
          image={list[activeAvatar - 1]}
          onAction={this.handleActiveClick}
          id="avatarPicker"
        />
        {visibleList && (
          <EnhancedListContainer
            list={list}
            activeAvatar={activeAvatar}
            loading={loading}
            handleAvatarChange={this.handleAvatarChange}
            handleActiveClick={this.handleActiveClick}
          />)}
      </div>
    );
  }
}

Picker.propTypes = {
  list: AvatarListShape.isRequired,
};

export default Picker;
