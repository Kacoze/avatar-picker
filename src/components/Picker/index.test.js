import React from 'react';
import renderer from 'react-test-renderer';
import { AvatarListExample } from '../../utils/testData';
import Picker from './index';

it('renders without crashing and match snapshoot', () => {
  const tree = renderer
    .create(<Picker list={AvatarListExample} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
