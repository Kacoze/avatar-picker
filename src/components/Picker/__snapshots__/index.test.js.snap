// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing and match snapshoot 1`] = `
<div
  className="Picker"
>
  <button
    className="avatarButton false"
    onClick={[Function]}
  >
    <img
      alt="Avatar 1"
      className="avatar"
      id="avatarPicker"
      src="images/avatar/avatar1.png"
    />
  </button>
</div>
`;
